### PR TITLE
[JDK21] Add support for JVMTI PopFrame

### PR DESCRIPTION
**Pull request by babsingh**
Previously, PopFrame returned JVMTI_ERROR_OPAQUE_FRAME for a virtual
thread.

In JDK21, PopFrame includes support for virtual threads as per the
JVMTI specification:
- Error if a virtual thread is not suspended and not the current
  thread.
- Error if a virtual thread is unomunted since it won't be able to
  pop the current frame.
- For a carrier thread with a virtual thread mounted, the details of
  the carrier thread are derived from targetThread->currentContinuation.

Related:
- #17715
- #17716

Also, there is no need to halt and resume a thread for inspection since
PopFrame expects the thread to be suspended as per the JVMTI spec. If a
thread is not suspended, it returns JVMTI_ERROR_THREAD_NOT_SUSPENDED.

### CODE REVIEW

**Comment by babsingh** [comment]
> Also, there is no need to halt and resume a thread for inspection since PopFrame expects the thread to be suspended as per the JVMTI spec. If a thread is not suspended, it returns JVMTI_ERROR_THREAD_NOT_SUSPENDED.

The above requirement to suspend is enforced from JDK8 to JDK21:

JDK8 - https://docs.oracle.com/javase/8/docs/platform/jvmti/jvmti.html#PopFrame
JDK21 - https://download.java.net/java/early_access/jdk21/docs/specs/jvmti.html#PopFrame

**Comment by babsingh** [comment]
Verified locally that [PopFrameTest](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/737f940cd67989a5eeb8dcb79c94a1d2d7266b27/test/hotspot/jtreg/serviceability/jvmti/vthread/PopFrameTest/PopFrameTest.java) passes with the changes from this PR.

@tajila Requesting your review.

File: runtime/jvmti/jvmtiStackFrame.c [hunk 1]
```diff
@@ -440,17 +440,43 @@ jvmtiPopFrame(jvmtiEnv* env,
 		ENSURE_CAPABILITY(env, can_pop_frame);
 
 		rc = getVMThread(
-				currentThread, thread, &targetThread, JVMTI_ERROR_OPAQUE_FRAME,
-				J9JVMTI_GETVMTHREAD_ERROR_ON_NULL_JTHREAD | J9JVMTI_GETVMTHREAD_ERROR_ON_DEAD_THREAD | J9JVMTI_GETVMTHREAD_ERROR_ON_VIRTUALTHREAD);
+				currentThread, thread, &targetThread, JVMTI_ERROR_NONE,
+				J9JVMTI_GETVMTHREAD_ERROR_ON_NULL_JTHREAD | J9JVMTI_GETVMTHREAD_ERROR_ON_DEAD_THREAD);
```
Comment (tajila) [inline runtime/jvmti/jvmtiStackFrame.c:444]: Is it legal to continue if its a dead thread?

File: runtime/jvmti/jvmtiStackFrame.c [hunk 2]
```diff
@@ -440,17 +440,43 @@ jvmtiPopFrame(jvmtiEnv* env,
...
+				rc = JVMTI_ERROR_OPAQUE_FRAME;
+				goto release;
+			}
+			threadObject = (NULL == thread) ? currentThread->threadObject : J9_JNI_UNWRAP_REFERENCE(thread);
+#endif /* JAVA_SPEC_VERSION >= 21 */
+
+			/* Error if the thread is not suspended and not the current thread. */
+			if ((currentThread != targetThread)
+#if JAVA_SPEC_VERSION >= 21
+			&& (0 == J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset))
```
Comment (tajila) [inline runtime/jvmti/jvmtiStackFrame.c:461]: should we be checking current thread or target thread?

File: runtime/jvmti/jvmtiStackFrame.c (hunk 1 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiStackFrame.c:444]: No. So, `JVMTI_ERROR_THREAD_NOT_ALIVE` will be returned by `getVMThread` when the `J9JVMTI_GETVMTHREAD_ERROR_ON_DEAD_THREAD` flag is set.

File: runtime/jvmti/jvmtiStackFrame.c (hunk 2 above)
Comment (babsingh) [reply runtime/jvmti/jvmtiStackFrame.c:461]: We need to check if the input/target `thread` is suspended. If `thread != currenThread` && `thread is not suspended`, then `JVMTI_ERROR_THREAD_NOT_SUSPENDED` is returned.

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
jenkins compile win jdk11

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Pull request merged**